File: src/traust_engine/data/secure-code-audit/opengrep-rules/python/data-exposure.py
import logging

logger = logging.getLogger(__name__)
log = logger


def startup(password, sasl_password, token_url, replica_count, headers):
    # ruleid: traust-python-data-exposure-secret-in-log
    logger.info("connecting with %s", password)

    # ruleid: traust-python-data-exposure-secret-in-log
    logging.debug("sasl config: %s", sasl_password)

    # ruleid: traust-python-data-exposure-secret-in-log
    print("pw:", password)

    # ruleid: traust-python-data-exposure-secret-in-log
    logger.error(f"auth failed for token {api_token}")

    # ruleid: traust-python-data-exposure-secret-in-log
    log.warning("db config %s", config["password"])

    # %-formatted operand naming a credential
    # ruleid: traust-python-data-exposure-secret-in-log
    log.info("access %s secret %s" % (access_key, secret_key))

    # ok: traust-python-data-exposure-secret-in-log
    logger.info("requesting token from %s", token_url)

    # ok: traust-python-data-exposure-secret-in-log
    logger.info("scaled to %d replicas", replica_count)

    # ok: traust-python-data-exposure-secret-in-log
    logger.error(f"request failed with headers count {len(headers)}")

    # ok: traust-python-data-exposure-secret-in-log
    logger.info("masked credential %s", mask(password))

    # ok: traust-python-data-exposure-secret-in-log
    logger.info("mounted secret %s", secret_name)


def credential_metadata(access_token_obj, masked_secret, token_length):
    # credential-adjacent metadata / pre-masked values (2026-07-29
    # dismissal classes) — identity and lifecycle facts, not the value
    # ok: traust-python-data-exposure-secret-in-log
    logger.debug("token length: %d", token_length)

    # ok: traust-python-data-exposure-secret-in-log
    logger.debug("token scopes: %s", access_token_obj.scopes)

    # ok: traust-python-data-exposure-secret-in-log
    logger.debug("token expiry: %s", access_token_obj.expires_at)

    # ok: traust-python-data-exposure-secret-in-log
    logger.debug("credentials: %s", masked_secret)

File: src/traust_engine/data/secure-code-audit/opengrep-rules/go/data-exposure.go
package rules

import (
	"fmt"
	"log"
	"os/exec"
)

// stand-in for k8s.io/klog/v2 so the fixture stays stdlib-only; the rules
// match on the call shape `klog.Infof(...)`, not the import.
var klog klogShim

type klogShim struct{}

func (klogShim) Infof(string, ...any) {}
func (klogShim) Info(...any)          {}

func logging(apiToken string, replicas int) {
	// ruleid: traust-go-data-exposure-secret-in-log
	log.Printf("connecting with token %s", apiToken)

	// ruleid: traust-go-data-exposure-secret-in-log
	klog.Infof("auth: %s", apiToken)

	// ruleid: traust-go-data-exposure-secret-in-log
	fmt.Printf("using %v", apiToken)

	// ok: traust-go-data-exposure-secret-in-log
	log.Printf("scaled to %d replicas", replicas)
}

func benignTokenMetadata(tokenURL string, secretName string) {
	// ok: traust-go-data-exposure-secret-in-log
	log.Printf("requesting token from %s", tokenURL)

	// ok: traust-go-data-exposure-secret-in-log
	log.Printf("mounted secret %s", secretName)
}

// credential-adjacent metadata classes from the 2026-07-29 dismissal
// corpus — identity/lifecycle facts about a credential, not its value
func benignCredentialMetadata(tokenExpiresOn string, tokenScope string, credentialFile string, secretErrors []error, tokenReview interface{}) {
	// ok: traust-go-data-exposure-secret-in-log
	log.Printf("token expires at %s", tokenExpiresOn)

	// ok: traust-go-data-exposure-secret-in-log
	log.Printf("token scope %s", tokenScope)

	// ok: traust-go-data-exposure-secret-in-log
	log.Printf("using credential file %s", credentialFile)

	// ok: traust-go-data-exposure-secret-in-log
	log.Printf("secret sync failures: %v", secretErrors)

	// ok: traust-go-data-exposure-secret-in-log
	log.Printf("token review outcome %v", tokenReview)
}

func argv(password string) {
	// ruleid: traust-go-data-exposure-cred-in-argv
	exec.Command("psql", "--password", password)

	// ruleid: traust-go-data-exposure-cred-in-argv
	exec.Command("vault", "login", "-token", password)

	// ok: traust-go-data-exposure-cred-in-argv
	exec.Command("psql", "--host", "db.internal")
}
